### PR TITLE
Adding functional tests for all training scenarios

### LINKED ISSUE #2906: Create functional tests for all Training scenarios
As laid out in #2498 , we need scenarios to cover the Training functionality we want fully supported in V1.

Scenarios
- I can provide multiple learners and easily compare evaluation metrics between them. 
- I can use an initial predictor to update/train the model for some trainers (e.g. linear learners like averaged perceptron). Specifically, start the weights for the model from the existing weights.  
- Metacomponents smartly restrict their use to compatible components.   Example: "When specifying what trainer OVA should use, a user will be able to specify any binary classifier. If they specify a regression or multi-class classifier ideally that should be a compile error."
- I can use OVA and easily add any binary classifier to it

### PR DESCRIPTION
**Pull request by rogancarr**
As laid out in #2498 , we need scenarios to cover the Training functionality we want fully supported in V1.

Scenarios

- I can provide multiple learners and easily compare evaluation metrics between them.
- I can use an initial predictor to update/train the model for some trainers (e.g. linear learners like averaged perceptron). Specifically, start the weights for the model from the existing weights.  
- Metacomponents smartly restrict their use to compatible components.   Example: "When specifying what trainer OVA should use, a user will be able to specify any binary classifier. If they specify a regression or multi-class classifier ideally that should be a compile error."
- I can use OVA and easily add any binary classifier to it

Fixes #2906

### CODE REVIEW

File: test/Microsoft.ML.Functional.Tests/Training.cs [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+        /// Training: It is easy to compare trainer evaluations on the same dataset.
+        /// </summary>
+        [Fact]
+        public void CompareTrainerEvaluations()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            // Get the dataset.
+            var data = mlContext.Data.LoadFromTextFile<TweetSentiment>(GetDataPath(TestDatasets.Sentiment.trainFilename),
+                separatorChar: TestDatasets.Sentiment.fileSeparator, hasHeader: TestDatasets.Sentiment.fileHasHeader,
```
Comment (sfilipi) [inline test/Microsoft.ML.Functional.Tests/Training.cs]: >hasHeader: TestDatasets.Sentiment.fileHasHeader,  [](start = 69, length = 49)

nit: newline this too.

File: test/Microsoft.ML.Functional.Tests/Training.cs [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+
+            // Create a selection of learners.
+            var sdcaTrainer = mlContext.BinaryClassification.Trainers.StochasticDualCoordinateAscent(
+                    new SdcaBinaryTrainer.Options { NumberOfThreads = 1 });
+
+            var fastTreeTrainer = mlContext.BinaryClassification.Trainers.FastTree(
+                    new FastTreeBinaryClassificationTrainer.Options { NumberOfThreads = 1 });
+
+            var ffmTrainer = mlContext.BinaryClassification.Trainers.FieldAwareFactorizationMachine(
+                    new FieldAwareFactorizationMachineBinaryClassificationTrainer.Options { });
```
Comment (sfilipi) [inline test/Microsoft.ML.Functional.Tests/Training.cs]: >new FieldAwareFactorizationMachineBinaryClassificationTrainer.Options { } [](start = 20, length = 73)

is this a required parameter? having to do this looks appalling :)

File: test/Microsoft.ML.Functional.Tests/Training.cs [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+                .Append(mlContext.MulticlassClassification.Trainers.OneVersusAll(anomalyDetectionTrainer))
+                .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel"));
+
+            // Fit the binary classification pipeline.
+            Assert.Throws<InvalidOperationException>(() => anomalyDetectionPipeline.Fit(data));
+
+            // Create a model training an OVA trainer with a binary classifier.
+            var binaryclassificationTrainer = mlContext.BinaryClassification.Trainers.LogisticRegression(
+                new LogisticRegression.Options { NumberOfIterations = 10, NumberOfThreads = 1, });
+            var binaryClassificationPipeline = mlContext.Transforms.Concatenate("Features", Iris.Features)
```
Comment (sfilipi) [inline test/Microsoft.ML.Functional.Tests/Training.cs]: >binaryClassificationPipeline [](start = 16, length = 28)

multiclassPipeline?

File: test/Microsoft.ML.Functional.Tests/Training.cs (hunk 2 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Training.cs]: Yes. We can't have an empty `()`. Filed #2927

---
In reply to: [264535384](https://github.com/dotnet/machinelearning/pull/2921#discussion_r264535384) [](ancestors = 264535384)

File: test/Microsoft.ML.Functional.Tests/Training.cs (hunk 3 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Training.cs]: Well, it's supposed be the OVA pipeline made up of binary classifiers.

---
In reply to: [264537263](https://github.com/dotnet/machinelearning/pull/2921#discussion_r264537263) [](ancestors = 264537263)

**Comment by artidoro** [comment]
Is this meant to be an exhaustive list of the trainers that can continue training starting from initial model parameter?
I believe there are some that don't allow training from initial model parameters do you intend to list them somewhere, or track them in an issue?

File: test/Microsoft.ML.Functional.Tests/Training.cs [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+            Common.AssertEqual(firstModelWeights.ToArray(), firstModelWeightsPrime.ToArray());
+            // Continued training should create a different model.
+            Common.AssertNotEqual(firstModelWeights.ToArray(), secondModelWeights.ToArray());
+        }
+
+        /// <summary>
+        /// Training: Models can be trained starting from an existing model.
+        /// </summary>
+        [Fact]
+        public void ContinueTrainingFieldAwareFactorizationMachines()
```
Comment (artidoro) [inline test/Microsoft.ML.Functional.Tests/Training.cs]: >ContinueTrainingFieldAwareFactorizationMachines [](start = 20, length = 47)

maybe remove the ending 's'

File: test/Microsoft.ML.Functional.Tests/Training.cs [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+                mlContext.Transforms.Concatenate("Features", Iris.Features)
+                    .AppendCacheCheckpoint(mlContext)
+                    .Append(mlContext.Transforms.Conversion.MapValueToKey("Label"))
+                    .Append(mlContext.MulticlassClassification.Trainers.OneVersusAll(kmeansTrainer))
+                    .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel")));
+
+            // Create a model training an OVA trainer with a multiclass classification trainer.
+            var multiclassTrainer = mlContext.MulticlassClassification.Trainers.LogisticRegression(
+                new MulticlassLogisticRegression.Options { MaximumNumberOfIterations = 10, NumberOfThreads = 1, });
+            Assert.Throws<ArgumentOutOfRangeException>(() =>
```
Comment (artidoro) [inline test/Microsoft.ML.Functional.Tests/Training.cs]: >ArgumentOutOfRangeException [](start = 26, length = 27)

So in the issue you mention that this should lead to a compile time error, ideally. Are we considering doing that?

**Comment by rogancarr** [comment]
I added tests for every trainer that allows a call to `Fit` with an initial model. You are right that most do not allow training from initial model parameters. Currently, we're not tracking these anywhere, and I know there are some issues from trainer to trainer with doing that. 

---
In reply to: [472508905](https://github.com/dotnet/machinelearning/pull/2921#issuecomment-472508905) [](ancestors = 472508905)

File: test/Microsoft.ML.Functional.Tests/Training.cs (hunk 5 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Training.cs]: A compile-time error is in our original spec. I'm not sure if we are planning on it, though. Let's discuss it in the issue and decide one way or another.

---
In reply to: [265232387](https://github.com/dotnet/machinelearning/pull/2921#discussion_r265232387) [](ancestors = 265232387)

Comment (artidoro) [reply test/Microsoft.ML.Functional.Tests/Training.cs]: Which issue?

Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Training.cs]: #2920

---
In reply to: [265307231](https://github.com/dotnet/machinelearning/pull/2921#discussion_r265307231) [](ancestors = 265307231)

**Comment by wschin** [comment]
Can OVA support FFM now? It could not back to the age of TLC.

**Comment by rogancarr** [comment]
@wschin Good question. It asks for a `ISingleFeaturePredictionTransformer<float>`, so it couldn't take an F**F**M, but it could take an FM. Shall we file an issue on it? Also, @ganik is working on constricting it to Binary Classifiers in https://github.com/dotnet/machinelearning/pull/2949.

**Pull request merged**